In [1]:
def funcion(x, y):
    return x**2 * (4 - 2.1*x**2 + x**4/3.0) + x*y + y**2 * (-4 + 4*y**2)

def busqueda_minimo(x_min, x_max, y_min, y_max, paso):
    mejor_x   = x_min
    mejor_y   = y_min
    mejor_val = funcion(x_min, y_min)

    x = x_min
    while x <= x_max:
        y = y_min
        while y <= y_max:
            val = funcion(x, y)
            if val < mejor_val:
                mejor_val = val
                mejor_x   = x
                mejor_y   = y
            y += paso
        x += paso

    return mejor_x, mejor_y, mejor_val

resultado_struct = busqueda_minimo(-2.0, 2.0, -1.0, 1.0, paso=0.001)

print("Programación estructurada:")
print(f"  x      = {resultado_struct[0]}")
print(f"  y      = {resultado_struct[1]}")
print(f"  f(x,y) = {resultado_struct[2]}")

Programación estructurada:
  x      = 0.08999999999989082
  y      = -0.7129999999999997
  f(x,y) = -1.0316274432090002


In [2]:
from random import random

def funcion(x, y):
    sum1 = x**2 * (4 - 2.1*x**2 + x**4/3.0)
    sum2 = x * y
    sum3 = y**2 * (-4 + 4*y**2)
    return sum1 + sum2 + sum3

def aleatorio(inf, sup):
    return random() * (sup - inf) + inf

class Particula:
    inercia   = 1.4
    cognitiva = 2.0
    social    = 2.0
    infx      = -2.0
    supx      =  2.0
    infy      = -1.0
    supy      =  1.0
    ajusteV   = 100.0

    def __init__(self):
        self.x        = aleatorio(Particula.infx, Particula.supx)
        self.y        = aleatorio(Particula.infy, Particula.supy)
        self.vx       = aleatorio(Particula.infx / Particula.ajusteV,
                                  Particula.supx / Particula.ajusteV)
        self.vy       = aleatorio(Particula.infy / Particula.ajusteV,
                                  Particula.supy / Particula.ajusteV)
        self.xLoc     = self.x
        self.yLoc     = self.y
        self.valorLoc = funcion(self.x, self.y)

    def actualizaVelocidad(self, xGlob, yGlob):
        cogX    = Particula.cognitiva * random() * (self.xLoc - self.x)
        socX    = Particula.social    * random() * (xGlob     - self.x)
        self.vx = Particula.inercia   * self.vx  + cogX + socX
        cogY    = Particula.cognitiva * random() * (self.yLoc - self.y)
        socY    = Particula.social    * random() * (yGlob     - self.y)
        self.vy = Particula.inercia   * self.vy  + cogY + socY

    def actualizaPosicion(self):
        self.x = self.x + self.vx
        self.y = self.y + self.vy
        self.x = max(self.x, Particula.infx)
        self.x = min(self.x, Particula.supx)
        self.y = max(self.y, Particula.infy)
        self.y = min(self.y, Particula.supy)
        valor = funcion(self.x, self.y)
        if valor < self.valorLoc:
            self.xLoc     = self.x
            self.yLoc     = self.y
            self.valorLoc = valor

def enjambreParticulas(particulas, iteraciones, reduccionInercia):
    mejorParticula = min(particulas, key=lambda p: p.valorLoc)
    xGlob          = mejorParticula.xLoc
    yGlob          = mejorParticula.yLoc
    valorGlob      = mejorParticula.valorLoc

    for iter in range(iteraciones):
        for p in particulas:
            p.actualizaVelocidad(xGlob, yGlob)
            p.actualizaPosicion()

        mejorParticula = min(particulas, key=lambda p: p.valorLoc)
        if mejorParticula.valorLoc < valorGlob:
            xGlob     = mejorParticula.xLoc
            yGlob     = mejorParticula.yLoc
            valorGlob = mejorParticula.valorLoc

        Particula.inercia *= reduccionInercia

    return xGlob, yGlob, valorGlob

nParticulas = 10
iteraciones = 100
redInercia  = 0.9

particulas    = [Particula() for i in range(nParticulas)]
resultado_pso = enjambreParticulas(particulas, iteraciones, redInercia)

print("Enjambre de partículas:")
print(f"  x      = {resultado_pso[0]}")
print(f"  y      = {resultado_pso[1]}")
print(f"  f(x,y) = {resultado_pso[2]}")

Enjambre de partículas:
  x      = -0.08984201388516055
  y      = 0.7126564047516615
  f(x,y) = -1.0316284534898774


# Comparacion de Resultados

Ambos métodos lograron encontrar soluciones muy cercanas al mínimo global de la función, aunque con enfoques distintos.

La programación estructurada evalúa el espacio de soluciones punto a punto con un paso fijo, lo que la hace determinista y confiable, pero su precisión depende directamente del tamaño del paso elegido. Con un paso de 0.001 encontró un valor de aproximadamente -1.0315, muy cercano al óptimo.

El enjambre de partículas, en cambio, exploró el espacio de forma dinámica guiando a las partículas con base en su experiencia individual y la del grupo. Esto le permitió converger a una solución ligeramente más precisa de -1.0316, sin necesidad de evaluar todos los puntos del dominio. Su principal desventaja es que, al usar aleatoriedad, los resultados pueden variar entre ejecuciones.